keps/sig-node/2712-pod-priority-based-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,544 @@
+    allocate different graceful shutdown durations for different pod priority
+    value ranges to terminate them gracefully during node shutdown
+
+
+### Implementation
+
+This implementation builds on top of the node graceful shutdown feature
+by introducing additional configuration. A new feature flag called
+`PodPriortityBasedGracefulShutdown` will be added to control the behavior
+of the kubelet.
+
+We will describe the configuration by using an example. Say, the
+following custom pod priority classes are created in a cluster:
+
+|Pod priority class name|Pod priority class value|
+|-----------------------|------------------------|
+|custom-class-a         | 100000                 |
+|custom-class-b         | 10000                  |
+|custom-class-c         | 1000                   |
+|regular/unset          | 0                      |
+
+We could set kubelet configuration to stop the pods as:
+
+|Pod priority class value|Shutdown period|
+|------------------------|---------------|
+| 100000                 |300 seconds    |
+| 10000                  |180 seconds    |
+| 1000                   |120 seconds    |
+| 0                      |60 seconds     |
+
+The above table implies that any pod with priority value >= 100000 will get
+300 seconds to stop, any pod with value >= 10000 and < 100000 will get 180
+seconds to stop, any pod with value >= 1000 and < 10000 will get 120 seconds to stop. 
+Finally, all other pods will get 60 seconds to stop.
+
+Note: We use priority values instead of names because k8s API design discourages
+using names and the values are more portable as well.
+
+One doesn't have to specify values corresponding to all of the classes. For
+e.g. the config could also be
+
+|Pod priority class value|Shutdown period|
+|------------------------|---------------|
+| 100000                 |300 seconds    |
+| 1000                   |120 seconds    |
+| 0                      |60 seconds     |
+
+
+In the above case, the pods with custom-class-b will go into the same bucket
+as custom-class-c for shutdown.
+
+If there are no pods in a particular range, then the kubelet will skip
+over the range entirely.
+
+If this feature is enabled and no configuration is provided, then no ordering
+action will be taken. The rationale is to allow some users to opt out of this
+if they are on a non-systemd distribution or have an older version of systemd
+with which this feature won't work.
+
+### Migration from the Node graceful shutdown feature
+
+If a user configures `ShutdownGracePeriod` to 5 minutes and `ShutdownGracePeriodCriticalPods`
+to 3 minutes, then it could be migrated to:
+
+|Pod priority class value|Shutdown period|
+|------------------------|---------------|
+| 1000000001             |180 seconds    |
+| 0                      |120 seconds    |
+
+Kubelet will be modified to only work with the config proposed in this KEP or the
+Node shutdown KEP. If both are specified, then it will be treated as a configuration
+error.
```

Comment:
Maybe add,

If neither are specified, then Graceful Node Shutdown feature is disabled.

Reply:
done